### PR TITLE
feat(134): enriched transfer receipts

### LINKED ISSUE #415: Transaction receipt 2nd pass: Transfers
## Story

**As a** user who placed an transfer
**I want** to see all the key details of my transfer transaction in my browser wallet
**So that** I can be certain I am approving the correct transaction

## Acceptance Criteria

- [ ] I can see a human readable 'receipt style' view of the key details when I am requested to approve / reject a Submit Order transaction
- [ ] The view contains the necessary data enrichment to provide
  - [ ] Size (with correctly formatted position decimals)
  - [ ] From Key
  - [ ] To Key
  - [ ] Indication of whether the transfers is between your own keys or an external one

## Tasks

- [x] Detailed sketch
- [x] Visual design
- [ ] Frontend tasks:
    - [ ] Implement UI
    - [ ] Integrate with backend
    - [ ] Write unit tests
- [ ] Backend tasks:
    - [ ] TBC
    - [ ] Write unit tests
- [ ] Write end to end tests
- [ ] Code review
- [ ] Team verify acceptance criteria met
- [ ] Exploratory QA testing 
- [ ] Capture bugs and new edge cases in tickets (this one or new TBD) 
- [ ] Write automated regression tests for the A/Cs (expanding on A/Cs where possible) 
- [ ] Add story and A/Cs to spec doc 
- [ ] Review with stakeholders
- [ ] Release

## Sketch

image

image

### PR DESCRIPTION
**Pull request by sam-keen**
Closes #415 

Screenshot 2023-09-18 at 15 09 01
Screenshot 2023-09-18 at 15 10 41

### CODE REVIEW

File: frontend/stores/assets-store.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  assets: null,
+  loading: false,
+  error: null,
+  async fetchAssets(request) {
+    try {
+      set({ loading: true, error: null })
+      const assets = await request(RpcMethods.Fetch, { path: 'api/v2/assets' })
+      set({ assets })
+    } catch (error) {
+      set({ error: 'Failed to fetch assets' })
```
Comment (dexturr) [inline frontend/stores/assets-store.ts:23]: You don't need this catch, request handles the error states. I wonder if we can make it handle loading states as well 🤔

File: frontend/stores/assets-store.spec.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import { RpcMethods } from '../lib/client-rpc-methods'
+import { useAssetsStore } from './assets-store'
+
+const assetsMock = [
```
Comment (dexturr) [inline frontend/stores/assets-store.spec.ts:4]: this mock is different to the data provided from the rest endpoint. Think this must have been generate through GQL.

See: https://vega-mainnet-data.commodum.io/api/v2/assets

File: frontend/stores/markets-store.spec.ts [hunk 3]
```diff
@@ -0,0 +1,315 @@
+import { RpcMethods } from '../lib/client-rpc-methods'
+import { useMarketsStore } from './markets-store'
+
+const marketsMock = [
```
Comment (dexturr) [inline frontend/stores/markets-store.spec.ts:4]: Same here, GQL data not REST

File: frontend/components/receipts/transfer/index.tsx [hunk 4]
```diff
@@ -18,24 +23,96 @@ const getTime = (transaction: Transaction) => {
...
+    )
+  }
+
+  return (
+    <ReceiptWrapper>
+      <h1 className="text-vega-dark-300">Amount</h1>
+
+      {assetsLoading || walletsLoading ? (
+        <div data-testid={locators.loading}>Loading...</div>
+      ) : error ? (
```
[the remark this answers is not in the document]
Comment (sam-keen) [reply frontend/components/receipts/transfer/index.tsx:110]: Sure. The reason I did this was that loading the basic view then replacing it meant that most of the time the view flashes up unpleasantly before being quickly replaced. I found displaying 'loading' to be nicer, but if you'd prefer for loading to be non-blocking that's fine

[the remark this answers is not in the document]
Comment (dexturr) [reply frontend/components/receipts/transfer/index.tsx]: Hmmm I wonder if we should do progressive enhancement with skeleton loaders. Out of scope for this PR I think, maybe we don't show a loading state for the first 50-100 milis or something.

File: frontend/stores/wallets.ts [hunk 5]
```diff
@@ -67,5 +71,22 @@ export const useWalletStore = create<WalletsStore>()((set, get) => ({
...
+            isOwnKey: true,
+            keyName: key.name
+          }
+        }
+      }
+    }
+    return {
+      isOwnKey: false,
+      keyName: undefined
+    }
```
Comment (dexturr) [inline frontend/stores/wallets.ts:90]: Just return the key, the presence of the key existing is proof that it is owned, also we should just return the whole key rather than a different object the adds no additional information.

Also you can simplify this:
```
return get().wallets.flatMap(w => w.keys).find(({ id }) => id === pubKey)
```

**Review by dalebennett1992 — CHANGES_REQUESTED** [review]
blocking for now given the tests indicate something is genuinely broken. About to investigate

**Comment by dalebennett1992** [comment]
ok so I get this simple through going through the onboarding flow

image


full error:
"Error: Failed to fetch data\n    at JSONRPCClient.onmessage (chrome-extension://jfaancmgehieoohdnmcdfdlkblfcehph/popup.js:106699:20)\n    at chrome-extension://jfaancmgehieoohdnmcdfdlkblfcehph/popup.js:106726:17"

File: frontend/components/receipts/transfer/enriched-transfer-view.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+export const locators = {
+  enrichedSection: 'enriched-section'
+}
+
+interface EnrichedTransferViewProps extends ReceiptComponentProps {
+  assetInfo: VegaAsset | undefined
+}
+
+export const EnrichedTransferView = ({ transaction, assetInfo }: EnrichedTransferViewProps) => {
+  const { getKeyInfo } = useWalletStore()
```
Comment (dexturr) [inline frontend/components/receipts/transfer/enriched-transfer-view.tsx:17]: Don't use zustand stores without an accessor, triggers a render on every state update
```suggestion
  const { getKeyInfo } = useWalletStore(state => ({ getKeyInfo: state.getKeyInfo }))
```

File: frontend/components/receipts/transfer/index.tsx [hunk 7]
```diff
@@ -1 +1,75 @@
...
+export const locators = {
+  whenSection: 'when-section',
+  whenElement: 'when-element',
+  loading: 'loading'
+}
+
+export const Transfer = ({ transaction }: ReceiptComponentProps) => {
+  const { loading: assetsLoading, assets, getAssetById } = useAssetsStore()
+  // We check whether wallets are loading as wallet data is used to enrich the transfer view
+  const { loading: walletsLoading } = useWalletStore()
```
Comment (dexturr) [inline frontend/components/receipts/transfer/index.tsx:34]: Same here

File: frontend/components/receipts/transfer/index.tsx [hunk 8]
```diff
@@ -1 +1,75 @@
...
+}
+
+export const locators = {
+  whenSection: 'when-section',
+  whenElement: 'when-element',
+  loading: 'loading'
+}
+
+export const Transfer = ({ transaction }: ReceiptComponentProps) => {
+  const { loading: assetsLoading, assets, getAssetById } = useAssetsStore()
```
Comment (dexturr) [inline frontend/components/receipts/transfer/index.tsx:32]: Same here

**Comment by dalebennett1992** [comment]
@dexturr you guys probably need to sort how this makes its way to main. This contains some of your untested changes right? Lots of code smells too. Give me a shout if you wanna chat about it

File: frontend/components/receipts/transfer/enriched-transfer-view.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+export const EnrichedTransferView = ({ transaction, assetInfo }: EnrichedTransferViewProps) => {
+  const { getKeyInfo } = useWalletStore((state) => ({ getKeyInfo: state.getKeyInfo }))
+  const { amount } = transaction.transfer
+  const decimals = Number(assetInfo?.details?.decimals)
+  const price = amount && decimals ? formatNumber(toBigNum(amount, decimals), decimals) : undefined
+  const symbol = assetInfo?.details?.symbol
+  const keyInfo = getKeyInfo(transaction.transfer.to)
+  const isOwnKey = keyInfo?.publicKey === transaction.transfer.to
```
Comment (dexturr) [inline frontend/components/receipts/transfer/enriched-transfer-view.tsx:23]: nitpick (non-blocking): just simplier, if it exists it is ours so can just do

```suggestion
  const isOwnKey = !!keyInfo
```

File: frontend/components/receipts/transfer/index.tsx [hunk 10]
```diff
@@ -1 +1,79 @@
...
+    assets,
+    getAssetById
+  } = useAssetsStore((state) => ({ loading: state.loading, assets: state.assets, getAssetById: state.getAssetById }))
+  // We check whether wallets are loading as wallet data is used to enrich the transfer view
+  const { loading: walletsLoading } = useWalletStore((state) => ({ loading: state.loading }))
+  const { asset } = transaction.transfer
+  const [assetInfo, setAssetInfo] = useState<VegaAsset | null>(null)
+
+  useEffect(() => {
+    if (assets.length > 0) {
```
Comment (dexturr) [inline frontend/components/receipts/transfer/index.tsx:43]: chore (blocking): I'd prefer to use the loading/error flags instead of deriving state that is based on implementation

File: frontend/components/receipts/transfer/index.tsx [hunk 11]
```diff
@@ -1 +1,79 @@
...
+      const res = getAssetById(asset)
+      if (res.id) {
+        setAssetInfo(res)
+      }
+    }
+  }, [assets, asset, getAssetById])
```
Comment (dexturr) [inline frontend/components/receipts/transfer/index.tsx:49]: chore (blocking): I'd move this logic into the enriched tx receipt view as this is a synchronous call and we only render that after the assets have loaded.

Also this way you need no useEffect either, just simply:

```
const assetInfo = getAssetById(asset)
```

File: frontend/components/receipts/transfer/index.tsx [hunk 12]
```diff
@@ -1 +1,79 @@
...
+
+  // Not supporting recurring transfers yet
+  if (transaction.transfer.recurring) return null
+  const time = getTime(transaction)
+
+  return (
+    <ReceiptWrapper>
+      <h1 className="text-vega-dark-300">Amount</h1>
+
+      {assetsLoading || walletsLoading || !assetInfo ? (
```
Comment (dexturr) [inline frontend/components/receipts/transfer/index.tsx:59]: follow up (non-blocking): out of scope for now, but error states should result in a basic view with a notification explaining what happened.

We should figure out how we can generalise this in some way 🤔

File: frontend/components/receipts/transfer/index.tsx [hunk 13]
```diff
@@ -1 +1,79 @@
...
+import { getDateTimeFormat } from '@vegaprotocol/utils'
+import { ReceiptWrapper } from '../utils/receipt-wrapper'
+import { useAssetsStore } from '../../../stores/assets-store.ts'
+import { useWalletStore } from '../../../stores/wallets.ts'
+import { VegaAsset } from '../../../types/rest-api.ts'
+
+const getTime = (transaction: Transaction) => {
+  const deliverOn = transaction.transfer.oneOff?.deliverOn
+  if (deliverOn) {
+    const deliverOnInSeconds = Math.floor(deliverOn / 1000)
```
Comment (dexturr) [inline frontend/components/receipts/transfer/index.tsx:17]: Nano to milli is 10^6 no?

```suggestion
    const deliverOnInSeconds = Math.floor(deliverOn / 1e6)
```

File: frontend/components/receipts/transfer/index.tsx (hunk 10 above)
Comment (dexturr) [reply frontend/components/receipts/transfer/index.tsx:43]: If we do the below this becomes redudant

File: test/e2e/extension-state.spec.ts [hunk 14]
```diff
@@ -41,6 +43,14 @@ describe('Check correct app state persists after closing the extension', () => {
     await driver.quit()
   })
 
+  beforeAll(async () => {
+    server.listen(test.test.mockPort)
+  })
+
+  afterAll(async () => {
+    await closeServerAndWait()
+  })
```
Comment (dexturr) [inline test/e2e/extension-state.spec.ts:52]: chore (blocking): spoke about with @dalebennett1992 on slack, but moving to a single test setup file

**Review by dexturr — APPROVED** [review]
